yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,683 @@
+FIXED_DA_GAS = 512
+```
+
+and
+
+```
+da_gas_per_field = DA_BYTES_PER_FIELD * DA_GAS_PER_BYTE
+
+note_hash_gas = da_gas_per_field * (number of notes)
+nullifier_gas = da_gas_per_field * (number of nullifiers)
+l2_to_l1_message_gas = da_gas_per_field * (number of l2_to_l1_messages)
+public_data_writes_gas = 2 * da_gas_per_field * (number of public_data_writes)
+```
+
+finally
+
+```
+log_gas = DA_GAS_PER_BYTE * (unencrypted_log_preimages_length + encrypted_log_preimages_length)
+teardown_da_gas = { specified by user }
+```
+
+:::note Deviation from ethereum calldata pricing
+
+Our `da_gas_used` will be correlated with the actual ethereum gas consumed by the published call, but not equal.
+
+This is because calculating the exact cost requires us to know how many non-zero bytes are in the calldata, which includes the `da_gas_used` itself. Since the `da_gas_used` must be produced in our kernel circuits, such a recursive calculation would unnecessarily complicate the circuit design.
+
+Instead, using a fixed cost plus scaling on actual resources consumed dramatically simplifies calculations, is a reasonable approximation, and affords abstraction if/when we transition to a different DA solution.
+:::
+
+
+# User transaction variables
+
+The user must be able to specify:
+
+- `da_gas_limit`: the **maximum** amount of DA gas the user is **willing** to pay for, **inclusive** of `teardown_da_gas`
+- `teardown_da_gas`: the amount of DA gas the user **will pay** for the teardown function
+- `max_fee_per_da_gas`: the maximum price the user is willing to pay per unit of DA gas in `$AZT`
+- `l2_gas_limit`: the **maximum** amount of l2 gas the user is **willing** to pay for, **inclusive** of `teardown_l2_gas`
+- `teardown_l2_gas`: the amount of l2 gas the user **will pay** for the teardown function
+- `max_fee_per_l2_gas`: the maximum price the user is willing to pay per unit of l2 gas in `$AZT`
+- `l1_gas_limit`: the **maximum** amount of l1 gas the user is **willing** to pay for, **inclusive** of `teardown_l1_gas`
+- `teardown_l1_gas`: the amount of l1 gas the user **will pay** for the teardown function
+- `max_fee_per_l1_gas`: the maximum price the user is willing to pay per unit of l1 gas in `$AZT`
+- `inclusion_fee`: the flat fee the user is willing to pay for transaction inclusion in `$AZT`
+
+
+
+# Sequencer checks
+
+Even though fees are not enforced by the protocol, the sequencer can check ahead of time if it is possible that the transaction **could** pay the fee.
+
+This is done by inspecting the public call enqueued for teardown, checking who the message sender is, and inspecting the balance of the `$AZT` for that sender.
+
+Note: it does not know by inspecting the enqueued call itself if the `pay_fee` function will be called, unless it has explicitly inspected/whitelisted the function ahead of time.
+
+However, it will know that if `pay_fee` is called, then message sender is the one that will pay the fee, and thus must have a balance greater than:
```

Comment:
> then message sender is the one that will pay the fee

This might not be a reliable indicator for whose balance will be paying the fee. The `msg_sender` of a private->public call could be a randomised value, to give the user privacy over which contract is making the call from private-land. The actual msg.sender that pays the fee could be some public contract that is enqueued, right? Or perhaps pay_fee might be invoked via a transfer_from call?
I might not be making sense, because I've forgotten how this part works.

Reply:
Presently the `pay_fee` function deducts the balance of `context.msg_sender`. Maybe this is not correct.